### PR TITLE
encoding/thrift: Add mechanisms to handle inbound requests via thriftrw's 'nowire' implementation

### PR DESCRIPTION
**Pull request by witriew**
thriftrw's 'nowire' implementation (known as 'streaming' in the thriftrw
context) is an alternative wire-compatible implementation of the Thrift
protocol that bypasses the intermediate 'thriftrw.Wire' representation and
'streams' the protocol directly to the related objects.

This provides the mechanisms to handle 'Call's made via the 'nowire'
implementation instead of the existing 'Wire'-intermediary format.  Doing so
involved creating a new type, `thriftNoWireHandler`, that implements both a
`transport.UnaryHandler` and a `transport.OnewayHandler`.

These call into a new interface, `NoWireHandler` (which will be implemented by
generated code, for each procedure), that understands how to unpack the request
using thriftrw's 'nowire' primitives (`stream.RequestReadeer`), execute the
request, and return objects that understand what's involved in sending back a
response, if it's a Unary call.

### CODE REVIEW

File: encoding/thrift/inbound_nowire.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+type NoWireHandler interface {
+	Handle(context.Context, *NoWireCall) (NoWireResponse, error)
+}
+
+func (t thriftNoWireHandler) Handle(ctx context.Context, treq *transport.Request, rw transport.ResponseWriter) error {
+	ctx, call := encodingapi.NewInboundCall(ctx)
+	res, err := doRequest(ctx, call, treq, rw, t.NoWireHandler, wire.Call, t.Protocol, t.Enveloping)
+	if err != nil {
```
Comment (AllenLuUber) [inline encoding/thrift/inbound_nowire.go:58]: missing something?

File: encoding/thrift/inbound_nowire.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		resp, err := noWireHandler.Handle(ctx, &nwc)
+		if err != nil {
+			return _emptyResponse, err
+		}
+
+		resp.ResponseWriter = responder
+		return resp, streamReader.ReadEnvelopeEnd()
+	}
+
+	responder := binary.NoEnvelopeResponder
```
Comment (AllenLuUber) [inline encoding/thrift/inbound_nowire.go:146]: move closer to where it is used L153 and inline due to single use. Same suggestion to L135
`resp.ResponseWriter = binary.NoEnvelopeResponder`

File: encoding/thrift/inbound_nowire.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	EnvelopeType  wire.EnvelopeType
+
+	StreamReader stream.Reader
+}
+
+type NoWireHandler interface {
+	Handle(context.Context, *NoWireCall) (NoWireResponse, error)
+}
+
+func (t thriftNoWireHandler) Handle(ctx context.Context, treq *transport.Request, rw transport.ResponseWriter) error {
```
Comment (AllenLuUber) [inline encoding/thrift/inbound_nowire.go:55]: can you leave a comment like, this is mostly identical to thriftUnaryHandler but skipping using the wire.Value from thriftrw? This helps with reviewers and maintenance.

File: encoding/thrift/inbound_nowire.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+		responder := binary.EnvelopeV1Responder{Name: eh.Name, SeqID: eh.SeqID}
+		nwc := NoWireCall{StreamReader: streamReader}
+		resp, err := noWireHandler.Handle(ctx, &nwc)
+		if err != nil {
+			return _emptyResponse, err
+		}
+
+		resp.ResponseWriter = responder
+		return resp, streamReader.ReadEnvelopeEnd()
```
Comment (AllenLuUber) [inline encoding/thrift/inbound_nowire.go:143]: move to helper? `decodeEnvelopedNoWireRequest`

File: encoding/thrift/inbound_nowire.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+	Reader        io.Reader
+	RequestReader stream.RequestReader
+	EnvelopeType  wire.EnvelopeType
+
+	StreamReader stream.Reader
+}
+
+type NoWireHandler interface {
+	Handle(context.Context, *NoWireCall) (NoWireResponse, error)
+}
```
Comment (Dogild) [inline encoding/thrift/inbound_nowire.go:53]: What is the reason we did not follow the same pattern as:
```
// UnaryHandler is a convenience type alias for functions that act as Handlers.
type UnaryHandler func(context.Context, wire.Value) (Response, error)

// OnewayHandler is a convenience type alias for functions that act as OnewayHandlers.
type OnewayHandler func(context.Context, wire.Value) error
```

I would prefer to have the same pattern here for consistency between the nowire and wire implementation (easy to maintain). It will also simplify the way the handler are being created in the thriftrw yarpc plugin and how it is register (from https://github.com/yarpc/yarpc-go/pull/2085)

File: encoding/thrift/inbound_nowire.go (hunk 4 above)
Comment (Dogild) [reply encoding/thrift/inbound_nowire.go:148]: +1 for this, let's have the same design/architecture as we have in inbound.go in the `decodeRequest` function

File: encoding/thrift/inbound_nowire.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+func (t thriftNoWireHandler) HandleOneway(ctx context.Context, treq *transport.Request) error {
+	ctx, call := encodingapi.NewInboundCall(ctx)
+	defer closeReader(treq.Body)
+
+	_, err := doRequest(ctx, call, treq, nil, t.NoWireHandler, wire.OneWay, t.Protocol, t.Enveloping)
+	return err
+}
+
+func doRequest(
```
Comment (Dogild) [inline encoding/thrift/inbound_nowire.go:101]: Let's call this `decodeNoWireRequest`. Please add same level of comment we have in `decodeRequest` for this function.

File: encoding/thrift/internal/observabilitytest/test/testserviceserver/server.go [hunk 7]
```diff
@@ -104,3 +107,50 @@ func (h handler) Call(ctx context.Context, body wire.Value) (thrift.Response, er
...
+		err error
+	)
+
+	if nwc.RequestReader != nil {
+
+		rw, err = nwc.RequestReader.ReadRequest(ctx, nwc.EnvelopeType, nwc.Reader, &args)
+
+	} else {
+		err = args.Decode(nwc.StreamReader)
+	}
```
Comment (Dogild) [inline encoding/thrift/internal/observabilitytest/test/testserviceserver/server.go:128]: Commenting on this generated code (I know the code was added in encoding/thrift/thriftrw-plugin-yarpc/server.go, but easier to understand what is going on).

Why do we have two ways for decoding the request here? How does it happen?

File: encoding/thrift/internal/observabilitytest/test/testserviceserver/server.go [hunk 8]
```diff
@@ -104,3 +107,50 @@ func (h handler) Call(ctx context.Context, body wire.Value) (thrift.Response, er
...
+		return thrift.NoWireResponse{}, yarpcerrors.InvalidArgumentErrorf(
+			"could not decode (via no wire) Thrift request for service 'TestService' procedure 'Call': %w", err)
+	}
+
+	success, appErr := h.impl.Call(ctx, args.Key)
+
+	hadError := appErr != nil
+	result, err := test.TestService_Call_Helper.WrapResponse(success, appErr)
+	var response thrift.NoWireResponse
+	response.ResponseWriter = rw
```
Comment (Dogild) [inline encoding/thrift/internal/observabilitytest/test/testserviceserver/server.go:140]: Is there anyway we could handle the ResponseWrite outside of this method? One of the big benefits we could get:
- We would not need `thrift.NoWireResponse`
- We could have a bunch of refactoring and code simplification by using the same piece of code to set the thrift.Response struct.

File: encoding/thrift/inbound_nowire.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+	Enveloping    bool
+}
+
+type NoWireCall struct {
+	Reader        io.Reader
+	RequestReader stream.RequestReader
+	EnvelopeType  wire.EnvelopeType
+
+	StreamReader stream.Reader
+}
```
Comment (Dogild) [inline encoding/thrift/inbound_nowire.go:49]: Why is this declared in yarpc and not in thriftrw? Should we no expect thriftrw to provide similar API for nowire and wire?

File: encoding/thrift/thriftrw-plugin-yarpc/server.go [hunk 10]
```diff
@@ -174,6 +175,69 @@ func (h handler) <.Name>(ctx <$context>.Context, body <$wire>.Value) (<$thrift>.
...
+			response.ApplicationErrorCode = extractor.YARPCErrorCode()
+		}
+		if appErr != nil {
+			response.ApplicationErrorDetails = appErr.Error()
+		}
+	}
+	return response, err
+	<end>
+}
+<end>
```
Comment (Dogild) [inline encoding/thrift/thriftrw-plugin-yarpc/server.go:240]: Please follow same structure as the wire implementation with the oneway and unary:

```
<if .OneWay>
// the entire Oneway implementation
<else>
// the entire Unary implementation
<end>
```

Having a bunch of if else in template is hard to follow and to maintain

File: encoding/thrift/internal/observabilitytest/test/testserviceserver/server.go (hunk 7 above)
Comment (witriew) [reply encoding/thrift/internal/observabilitytest/test/testserviceserver/server.go:128]: This happens because the caller of this `Handle` may not have a `proto` that implements a `stream.RequestReader`.  If the caller has a `proto` that implements `stream.RequestReader`, then the logic is quite simple since at that call site, it simply returns that request:
```
	if reqReader, ok := proto.(stream.RequestReader); ok {
		nwc := NoWireCall{
			Reader:        treq.Body,
			RequestReader: reqReader,
			EnvelopeType:  reqEnvelopeType,
		}
		return noWireHandler.Handle(ctx, &nwc)
	}
```

The differences between this and the Wire utilization, is that even outside of being able to make the request to `Handle`, `Read...` operations are still needed once the request is processed.

An argument can be made that will wrap a `stream.Protocol` into a struct that implements a `stream.RequestReader` which should simplify this.   That should make this a lot cleaner - will implement that.

File: encoding/thrift/internal/observabilitytest/test/testserviceserver/server.go (hunk 8 above)
Comment (witriew) [reply encoding/thrift/internal/observabilitytest/test/testserviceserver/server.go:140]: The response write isn't actually happening here.  The object that gets returned understands what to use to perform the write, but doesn't do it; the write happens in `inbound_nowire.go` in the `Handle` method.

We still wouldn't be able to use the same `thrift.Response` struct because its internals are based on ThriftRW's Wire representation (and its respective `envelope.Enveloper` which contains the response body), where the NoWire representation needs a writer as well as specific response body (that implements `stream.Enveloper`).

File: encoding/thrift/inbound_nowire.go (hunk 4 above)
Comment (witriew) [reply encoding/thrift/inbound_nowire.go]: With the code updated to resolve https://github.com/yarpc/yarpc-go/pull/2084#discussion_r682968284, this ended up being a lot cleaner.  Let me know if having a helper there is still preferred.

File: encoding/thrift/inbound_nowire.go [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+	}
+
+	return nil
+}
+
+func (t thriftNoWireHandler) HandleOneway(ctx context.Context, treq *transport.Request) error {
+	ctx, call := encodingapi.NewInboundCall(ctx)
+	defer closeReader(treq.Body)
+
+	_, err := decodeNoWireRequest(ctx, call, treq, nil, t.NoWireHandler, wire.OneWay, t.Protocol, t.Enveloping)
```
Comment (abhinav) [inline encoding/thrift/inbound_nowire.go:116]: That's a lot of shared parameters that are just fields of t. I would make this a method of t.

Fixing.

File: encoding/thrift/inbound_nowire.go [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+// except that thriftNoWireHandler implements both transport.UnaryHandler and
+// transport.OnewayHandler through a single type, utilizing the "nowire"
+// (streaming in the ThriftRW context, bypassing the intermediate "Wire"
+// representation) implementation.
+//
+// The UnaryHandler and OnewayHandler implementations then call into type that
+// implements NoWireHandler, which understands how to unpack and invoke the
+// request, given the ThriftRW primitives for reading the raw representation.
+type thriftNoWireHandler struct {
+	NoWireHandler NoWireHandler
```
Comment (abhinav) [inline encoding/thrift/inbound_nowire.go:47]: We're inside a struct called thriftNoWireHandler, so we don't need to name the field NoWireHandler. It can just be "Handler" or similar.

File: encoding/thrift/inbound_nowire.go [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+		}
+	}
+	return noWireHandler.Handle(ctx, &nwc)
+}
+
+// reqReaderProto is an implementation of ThriftRW's stream.RequestReader in
+// case the provided stream.Protocol does not implement the necessary
+// `ReadRequest` to discover the correct enveloping.
+type reqReaderProto struct {
+	stream.Protocol
```
Comment (abhinav) [inline encoding/thrift/inbound_nowire.go:163]: unnecessary embedding IMO.

Fixing.

File: encoding/thrift/inbound_nowire.go [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+func (p *reqReaderProto) ReadRequest(
+	ctx context.Context,
+	et wire.EnvelopeType,
+	r io.Reader,
+	body stream.BodyReader,
+) (stream.ResponseWriter, error) {
+	sr := p.Protocol.Reader(r)
+	defer sr.Close()
+
+	if p.enveloping {
```
Comment (abhinav) [inline encoding/thrift/inbound_nowire.go:180]: I'm not a big fan of two code paths that are all but completely different are inside the same type with a bool switch. The switch should be in decodeNoWireResponse, constructing either an enveloped variant of this type, or a non-enveloped.

Fixing.

**Comment by abhinav** [comment]
Some context on the discussion @witriew and I had about this, which, in
hindsight, I should have written down and shared with you, @Dogild and
@AllenLuUber.

Some of the decisions made in encoding/thrift, after a couple years of
hindsight, were not great. For example, one of these is the decision to use
newtype wrappers around `func`s (like `type UnaryHandler func(..)`). This makes
it impossible to have a clean upgrade path for that implementation. For
example, had we used a singular `Handler` interface, we'd be able to add the
new `NoWireHandler` interface and internally upgrade from `Handler` to
`NoWireHandler` if available with fewer intrusive changes.

When Wit was designing the interfaces here, I suggested that since these were
new APIs largely unencumbered by the burden of backwards comaptibility, he
should err towards choices that we would make today if we were designing these
APIs from scratch.

File: encoding/thrift/inbound_nowire_test.go [hunk 15]
```diff
@@ -0,0 +1,374 @@
...
+type responseEnveloper struct {
+	name         string
+	envelopeType wire.EnvelopeType
+}
+
+var _ stream.Enveloper = (*responseEnveloper)(nil)
+
+func (re responseEnveloper) MethodName() string              { return re.name }
+func (re responseEnveloper) EnvelopeType() wire.EnvelopeType { return re.envelopeType }
+func (re responseEnveloper) Encode(stream.Writer) error      { return nil }
```
Comment (Dogild) [inline encoding/thrift/inbound_nowire_test.go:61]: Why are we implementing a stub for `stream.Enveloper` instead of using `MockEnveloper` from streamtest?

File: encoding/thrift/inbound_nowire_test.go [hunk 16]
```diff
@@ -0,0 +1,374 @@
...
+
+type responseWriter struct{ err error }
+
+var _ stream.ResponseWriter = (*responseWriter)(nil)
+
+func (rw responseWriter) WriteResponse(wire.EnvelopeType, io.Writer, stream.Enveloper) error {
+	return rw.err
+}
```
Comment (Dogild) [inline encoding/thrift/inbound_nowire_test.go:69]: Why not using `MockResponseWriter`?

File: encoding/thrift/inbound_nowire_test.go [hunk 17]
```diff
@@ -0,0 +1,374 @@
...
+
+type responseHandler struct {
+	t   *testing.T
+	nwc *NoWireCall
+
+	reqBody stream.BodyReader
+	body    stream.Enveloper
+}
+
+var _ NoWireHandler = (*responseHandler)(nil)
```
Comment (Dogild) [inline encoding/thrift/inbound_nowire_test.go:79]: Is there any chance to use a mock for NoWireHandler?

**Comment by abhinav** [comment]
Okay, so this stack isn't green because of:

```
# go.uber.org/yarpc/encoding/thrift [go.uber.org/yarpc/encoding/thrift.test]
./inbound_nowire_test.go:248:3: cannot use proto (type *streamtest.MockRequestReader) as type stream.Protocol in field value:
        *streamtest.MockRequestReader does not implement stream.Protocol (missing Reader method)

```

That's a result of https://github.com/thriftrw/thriftrw-go/pull/532 — RequestReaders and Protocols are different interfaces now so the same object cannot be used for them.

Given that this is a new code path, I don't see value in doing upgrade-protocol-to-requestreader dance and just having it be RequestReader-only: All stream based implementations must implement RequestReader. Going to update the PR stack with fixes.

**Review by abhinav — APPROVED** [review]
LGTM besides nits.

File: encoding/thrift/inbound_nowire_test.go (hunk 15 above)
Comment (Dogild) [reply encoding/thrift/inbound_nowire_test.go:61]: I would prefer the Mock for the following reasons:
- Yes readability is a bit harder, but mock are known by most of the developers, no need of understand what the stub is doing.
- More control IMO, for instance you could do `env.EXPECT().MethodName().Return("foo").Times(1)`. This IMO adds more control on what the test should actually test.

File: encoding/thrift/thriftrw-plugin-yarpc/server.go [hunk 18]
```diff
@@ -86,6 +86,7 @@ func New(impl Interface, opts ...<$thrift>.RegisterOption) []<$transport>.Proced
 					Type: <$transport>.Unary,
 					Unary: <import $unaryWrapperImport>.<$unaryWrapperFunc>(h.<.Name>),
 				<end>
+					NoWire: <.Name>_NoWireHandler{impl},
```
[the remark this answers is not in the document]
Comment (Dogild) [reply encoding/thrift/thriftrw-plugin-yarpc/server.go:89]: +1

File: encoding/thrift/internal/observabilitytest/test/testserviceserver/server.go [hunk 19]
```diff
@@ -57,6 +58,8 @@ func New(impl Interface, opts ...thrift.RegisterOption) []transport.Procedure {
 
 					Type:  transport.Unary,
 					Unary: thrift.UnaryHandler(h.Call),
+
+					NoWire: Call_NoWireHandler{impl},
```
[the remark this answers is not in the document]
Comment (Dogild) [reply encoding/thrift/internal/observabilitytest/test/testserviceserver/server.go:62]: See @abhinav comment go using `<>` and `-`

File: encoding/thrift/internal/observabilitytest/test/testserviceserver/server.go [hunk 20]
```diff
@@ -104,3 +107,44 @@ func (h handler) Call(ctx context.Context, body wire.Value) (thrift.Response, er
 
 	return response, err
 }
+
+type Call_NoWireHandler struct{ impl Interface }
```
[the remark this answers is not in the document]
Comment (Dogild) [reply encoding/thrift/internal/observabilitytest/test/testserviceserver/server.go:111]: True, what about using `template.FuncMap` to get lowercase (see https://stackoverflow.com/questions/21031108/golang-template-use-pipe-to-uppercase-string)? This method does not need to be public

File: encoding/thrift/inbound_nowire.go [hunk 21]
```diff
@@ -0,0 +1,228 @@
...
+
+	return binary.EnvelopeV1Responder{
+		Name:  eh.Name,
+		SeqID: eh.SeqID,
+	}, nil
+}
+
+// noEnvelopeReader implements ThriftRW's stream.RequestReader, decoding
+// requests without enveloping support.
+type noEnvelopeReader struct {
```
[the remark this answers is not in the document]
Comment (witriew) [reply encoding/thrift/inbound_nowire.go]: There aren't individual structs for different enveloping anymore.  Only envelope detection is used.

File: encoding/thrift/inbound_nowire_test.go (hunk 17 above)
Comment (witriew) [reply encoding/thrift/inbound_nowire_test.go:80]: In the case of the `NoWireHandler` stub, I wanted to ensure that the stuff thats passed into `HandleNoWire` is what is expected.  Since all of the calls here have effectively the same checks, it seems cleaner to have the stub that does all of this at once instead of repeating these checks for each of the mocks.

File: encoding/thrift/inbound_nowire_test.go (hunk 15 above)
Comment (witriew) [reply encoding/thrift/inbound_nowire_test.go]: Done.

File: encoding/thrift/inbound_nowire.go [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+// NoWireCall contains all of the required objects needed for an underlying
+// Handle needs to unpack any given request.
+type NoWireCall struct {
+	Reader        io.Reader
+	RequestReader stream.RequestReader
+	EnvelopeType  wire.EnvelopeType
+}
+
+// NoWireHandler is implemented by the generated code for each method that the
+// server needs to implement.  It is responsible for unpacking the request,
```
Comment (Dogild) [inline encoding/thrift/inbound_nowire.go:46]: nit: extra space before the It

**Review by Dogild — APPROVED** [review]
